pkg/util/watch/README.md

```diff
@@ -0,0 +1,95 @@
+# `util/watch` explainer
+
+Here's a high-level overview of `.../util/watch` - our abstraction for the Kubernetes Watch API.
+
+Our implementation is similar to the [`client-go` watch implementation], with some notable
+differences:
+
+1. client-go's is type-erased; ours is generic
+2. client-go's periodically resyncs when necessary, but does not provide the external control or
+   ordering guarantees necessary to use it. Ours provides explicit control for relisting, with
+   notifications when that relisting is complete.
+3. Ours has some additional features, like support for custom indexes
+4. Ours has better observability, by exposing metrics that describe both the API calls and their
+   results, and the current state of the watch (e.g. whether it's healthy).
+
+[`client-go` watch implementation]: https://pkg.go.dev/k8s.io/client-go/tools/watch
+
+---
+
+## How watching generally works
+
+At a high level, watching _generally_ works by first calling `List` to fetch the current state of
+the items, and then repeatedly calling `Watch` _with the [resource version] from listing_ to stream
+the changes to the objects since the initial fetch. Sometimes there are errors with watching that
+require calling `List` again (e.g. resource version is too old).
+
+[resource version]: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
+
+So, our general flow looks like:
+
+```
+for item in List() {
+    update store with item
+}
+stream := Watch()
+loop {
+    // stream events...
+    for item in stream {
+        if item is error {
+            goto relist
+        }
+
+        update store with item
+    }
+    goto rewatch
```

Comment:
Partially clarified this with 7ed7c74b765b4c22f3c4b979f210132768a3a522 — the `goto rewatch` occurs when we see the channel is closed.

When the channel is closed (IIUC it can periodically happen even if the stream is otherwise healthy), AFAIK the correct thing to do is to try to reconnect

Does that match your understanding?